src/python/pants/engine/query.py

```diff
@@ -0,0 +1,240 @@
+        property, the .parse_from_args() method is invoked with `<args...>` (shlexed, so split by
+        spaces).
+        """
+
+    @classmethod
+    @abstractmethod
+    def parse_from_args(cls, *args):
+        """Create an instance of this class from variadic positional string arguments.
+
+        This method should raise an error if the args are incorrect or invalid.
+        """
+
+
+@dataclass(frozen=True)
+class QueryAddresses:
+    addresses: Addresses
+
+
+@dataclass(frozen=True)
+class OwnerOf(QueryComponent):
+    files: Tuple[str]
+
+    function_name = 'owner_of'
+
+    @classmethod
+    def parse_from_args(cls, *args):
+        return cls(files=tuple([str(f) for f in args]))
+
+
+@rule
+async def owner_of_request(owner_of: OwnerOf) -> QueryAddresses:
+    request = OwnersRequest(sources=owner_of.files)
+    owners = await Get[Owners](OwnersRequest, request)
+    return QueryAddresses(Addresses(bfa.to_address() for bfa in owners.addresses))
+
+
+@dataclass(frozen=True)
+class ChangesSince(QueryComponent):
+    changes_since: str
+    include_dependees: IncludeDependeesOption
+
+    function_name = 'changes_since'
+
+    @classmethod
+    def parse_from_args(cls, changes_since, include_dependees=IncludeDependeesOption.NONE):
+        return cls(changes_since=str(changes_since),
+                             include_dependees=IncludeDependeesOption(include_dependees))
+
+
+@rule
+async def changes_since_request(
+        scm_wrapper: UncachedScmWrapper,
+        changes_since: ChangesSince,
+) -> QueryAddresses:
+    scm = scm_wrapper.scm
+    changed_options = ChangedOptions(
+        changes_since=changes_since.changes_since,
+        diffspec=None,
+        include_dependees=changes_since.include_dependees,
+        fast=True,
+    )
+    changed = await Get[ChangedAddresses](ChangedRequest(
+        sources=tuple(changed_options.changed_files(scm=scm)),
+        include_dependees=changed_options.include_dependees,
+    ))
+    return QueryAddresses(changed.addresses)
+
+
+@dataclass(frozen=True)
+class ChangesForDiffspec(QueryComponent):
+    diffspec: str
+    include_dependees: IncludeDependeesOption
+
+    function_name = 'changes_for_diffspec'
+
+    @classmethod
+    def parse_from_args(cls, diffspec, include_dependees=IncludeDependeesOption.NONE):
+        return cls(diffspec=str(diffspec),
+                             include_dependees=IncludeDependeesOption(include_dependees))
+
+
+@rule
+async def changes_for_diffspec_request(
+        scm_wrapper: UncachedScmWrapper,
+        changes_for_diffspec: ChangesForDiffspec,
+) -> QueryAddresses:
+    scm = scm_wrapper.scm
+    changed_options = ChangedOptions(
+        changes_since=None,
+        diffspec=changes_for_diffspec.diffspec,
+        include_dependees=changes_for_diffspec.include_dependees,
+        fast=True,
+    )
+    changed = await Get[ChangedAddresses](ChangedRequest(
+        sources=tuple(changed_options.changed_files(scm=scm)),
```

Comment:
I think that the `UncacheableSCM` wrapper is sufficient to avoid blocking this change, but it would be good to reference #9015 in a TODO nearby it. I think the next step to making that more rigorous be to expose an uncacheable, absolute process execution API that would allow for invoking `git` correctly.

But on the other hand, if these lines were moved to where you currently construct the SCM wrapper, I don't think it would be necessary to actually address this issue yet? Essentially, the computed changed files would be pushed down as a root, rather than the SCM itself.

Reply:
Resolved with `@_uncacheable_rule` with a big TODO.